### DISCUSSION TITLE
Comparison with other software

### DISCUSSION THREAD

**Discussion by connorferster** [robbievanleeuwen/concrete-properties#9] [General]
Hi @robbievanleeuwen!

I have been playing with `concreteproperties` and just created this gist:
https://gist.github.com/connorferster/7fe8c1687dab86d0d84be4ee0287515a

The comparison with the geometric solver used in spColumn is quite good! 

The interesting thing I noticed is that, as the axial load increases, the divergence between the two solvers also increases. Do you have any thoughts on why this might be?

The differences are so small (and modelling a concrete section is inherently without high accuracy) so I do not think this is a big deal at all. I was very happy I was able to match the results of this commercial solver!

I have said it before and imma keep saying it: `concreteproperties` is just awesome.

**Comment by robbievanleeuwen** [comment]
Gday @connorferster!

Thanks for performing the comparison! Glad to hear the results are looking good!

Not too sure on the small differences? Can't say I'm totally on top of how spColumn works, but in the Australian code the capacity factor ``phi`` varies depending on the axial load, could it be related to this? Or perhaps you've controlled this in spColumn so it's a like for like comparison. Other thoughts - does spColumn deduct steel area from concrete area, *concreteproperties* does so maybe it's this? This could explain the slightly deeper neutral axis under higher axial load? For this simple geometry, a hand calc could verify what the correct neutral axis is for this axial load. I can have a look at this tomorrow :)

**Reply by connorferster** [reply]
Hi @robbievanleeuwen, thanks for the ideas on this. I will look into how concrete area is calculated with spColumn and will read your AS3600 implementation to see how the phi factor adjusts with axial load. Will report back on the area!

**Reply by robbievanleeuwen** [reply]
Re: the code, is your spColumn run providing results from a design code or just the constant 0.65 factor for concrete and 0.85 factor for steel as per your gist? FYI in the most recent version of AS 3600 there is this cumbersome calculation for the capacity reduction factor, which gets applied after the nominal capacity is found (using unfactored material properties).

<img width="695" alt="Screen Shot 2022-07-21 at 12 02 13 am" src="https://user-images.githubusercontent.com/18841910/180001780-1e9b7d6d-5914-482a-88cb-2b4f298202e8.png">

``N_ub`` is the axial force at the balanced point.

**Reply by robbievanleeuwen** [reply]
@connorferster and @Agent6-6-6, I've attached a hand calculation veryifying the neutral axis depth result from *concreteproperties* for ``N = 3000`` kN.

[2022_07_21 8_18 pm Office Lens.pdf](https://github.com/robbievanleeuwen/concrete-properties/files/9158498/2022_07_21.8_18.pm.Office.Lens.pdf)

The calculated neutral axis depth gives a net axial force of 3000 kN so this should validate the *concreteproperties* result. Might be worth you two having a quick look over this to verify but I think it's good. It thus follows that the spColumn neutral axis depth would give a different net axial force (not equal to 3000 kN) as there is only one unique neutral axis depth for a given axial force.

Perhaps there is something else going on in the background that spColumn is doing that we don't know about? Let me know your thoughts!

**Reply by robbievanleeuwen** [reply]
Also thought it would be worth mentioning that the ``calculate_ultimate_stress()`` method here gives us a *look under the hood* at the *concreteproperties* results.

For example continuing on from your gist:
```python
from rich.pretty import pprint

stress_res = conc_sec.calculate_ultimate_stress(res_3000)

# print stress results
pprint(stress_res.concrete_forces)
pprint(stress_res.steel_strains)
pprint(stress_res.steel_stresses)
pprint(stress_res.steel_forces)
```

**Output:**
*concrete_forces* - first item in list is tensile region (no force), second item is the compression zone *(force, lever arm)*
```console
[(0.0, 0, 0), (2465480.0831598416, 145.8069943062731, 0)]
```

*steel_strains*
```console
[
│   0.002618250851468556,
│   0.002618250851468557,
│   0.002618250851468556,
│   0.002618250851468557,
│   0.002618250851468556,
│   -0.0009364001188475868,
│   -0.0009364001188475868,
│   -0.0009364001188475868,
│   -0.0009364001188475868,
│   -0.0009364001188475868
]
```

*steel_stresses*
```console
[
│   array(340.),
│   array(340.),
│   array(340.),
│   array(340.),
│   array(340.),
│   array(-187.28002377),
│   array(-187.28002377),
│   array(-187.28002377),
│   array(-187.28002377),
│   array(-187.28002377)
]
```

*steel_forces* - *(force, lever arm)*
```console
[
│   (238000.0000000075, 196.92947450938487, 0),
│   (238000.00000000742, 196.92947450938493, 0),
│   (238000.00000000742, 196.92947450938487, 0),
│   (238000.00000000742, 196.92947450938493, 0),
│   (238000.00000000742, 196.92947450938487, 0),
│   (-131096.01663866633, -70.43052549061508, 0),
│   (-131096.0166386663, -70.43052549061508, 0),
│   (-131096.0166386663, -70.43052549061508, 0),
│   (-131096.0166386663, -70.43052549061508, 0),
│   (-131096.0166386663, -70.43052549061508, 0)
]
```

The forces above match my hand calcs 👍 

More on the ``StressResult`` object [here](https://robbievanleeuwen.github.io/concrete-properties/rst/results.html#stress-analysis).

**Reply by connorferster** [reply]
I am completely on-board @robbievanleeuwen. I did a hand calc on the spColumn results today at lunch and indeed found that they were _not_ in equilibrium. Further, unlike `concreteproperties`, you have limited means of investigating what is happening in the calculation to generate the result with spColumn. Closed-source software strikes again.

The `StressResult` object is a really great design decision. I love it.

**Reply by Agent6-6-6** [reply]
@connorferster It could be a little out by the fact you've used the default concrete stress block modifiers (alpha and beta factors within ACI318) which obviously have more precision within spColumn than the 4 decimal places used to compare with? 

In SPCOLUMN you can manually define the strength reduction factors and the alpha and beta factors, just use whatever value you like if you want to compare with a similarly defined value in concreteproperties. I think from memory the strength reduction factors can be defined under confinement as user defined, and the alpha and beta factor via the material definition from memory.

You can also output the entire interaction surface from spColumn to a text file to compare more than two axial load points to see if there is a general trend or if it diverges further at certain values.

**Reply by Agent6-6-6** [reply]
@connorferster, @robbievanleeuwen 
I'd like to start another conversation on these differences, having spent some time now creating tests for the NZS3101 code implementation based on independent assessment of moment capacity and neutral axis depths (based on spreadsheet calculations) based on the following assumptions I see some (small) differences in the answers that cause the tests to fail:-

- independent hand calculations are typically based on assumption that impact of bar and hence concrete displaced by bar is only considered once the compression block depth goes below the bar centreline.
- this is considered the 'normal way' to analyse a section by hand with compression reinforcement.  By effectively considering the compression block without the bar holes, and working with a net bar stress of = $f_y-\\alpha f'_c$ in the bars within the compression block (if they were yielding) or similar approach based on assessing strains if not yielding to determine actual bar stress if lower than yield.
-  However concrete_properties analyses the compression block I believe with the holes being present, but then only considers the bar once the compression block is past the bar centreline. But considers the concrete hole before getting to the centre of the bar. This seems like an inconsistent assumption.
- So this leads to a situation where say you're 1/3 of the way into the bar (using non meshed reinforcement), so the bar is not considered in compression, but the reduction in bar area from the concrete is starting to be considered in compression before the consideration of the bar. This is what I believe leads to the differences between a 'hand analysis' and a 'concreteproperties analysis'. This does not seem like a consistent approach with a "typical way things are done" approach for cross section analysis.

It's hard to generalise the difference (as a percentage it is probably small obviously), but it leads to the question of what values to put in the tests and how accurate does the analysis need to be for comparison to known examples. It could be altered to utilise the net stress approach I'd imagine which would seem more correct and consistent with typical concrete analysis methods.

This makes me question whether directly considering the bar holes in the concrete is really the correct approach as you're departing from the assumptions made in every analysis textbook for example.

Thoughts?

**Reply by connorferster** [reply]
@Agent6-6-6 

That's interesting. I went back to review the original Gist I posted (up top). In the Gist, the neutral axis seems well into the middle-ish part of the section so, I think, the holes in the concrete are being compensated with by the bar being fully in the compression zone, correct? Do you think the holes-in-concrete would create the effect I saw of the results comparison having a greater $\Delta$ when the compression load is higher?

**Reply by Agent6-6-6** [reply]
Yeah need to look into it further to really comment, and compare both methods in some independent calculations in Excel. At the moment I'm only comparing my spreadsheet using the 'normal calculation method' directly with concreteproperties answers and just noting some differences and coming up with a reasonable explanation (which as far as I can tell at this stage is the only logical way I can attribute the differences to).

> Do you think the holes-in-concrete would create the effect I saw of the results comparison having a greater  when the compression load is higher?

Possibly, my guess at this stage without really corroborating it is that if you have the neutral axis depth within any bars and depending on the bar geometry it leads to slight differences in neutral axis depths. Does not explain your differences though if your neutral axis depth is well below the top reinforcement within a region where full design compression stress in concrete is being achieved.

Possibly related or in support of this theory is if I alter `n_points` for the bars and holes, I get marginally different neutral axis depths which leads me towards thinking the defined bar holes are potentially leading to the differences in results. Even at larger neutral axis depths where this should not be an issue, I find changing `n_points` from say 4 to 500 changes the moment capacity ever so slightly (like 1st/2nd decimal place being different with zero axial load).

**Reply by Agent6-6-6** [reply]
@connorferster, checked some cases with higher axial loads and uniaxial bending to NZS3101 code and I see the same trend, whereby concreteproperties capacity is tending towards overestimating the capacity by up to 3% compared with traditional methods of estimating the capacity of a cross section utilising effective net bar stresses. So on say a 2000kNm capacity, this means up to 60kNm discrepancy which is enough to create arguments amongst engineers!

I've also checked similar sections in SPColumn and I get within 0.05% typically between SPColumn and my hand checks which is sort of the agreement I'd typically expect when comparing NZS3101 results.

**Reply by Agent6-6-6** [reply]
@connorferster, I also chucked in your Canadian section from your gist into my spreadsheet just to confirm it wasn't my spreadsheet and I get agreement with the concreteproperties results to within fractions of a percent for both 1500kN and 3000kN (and agreement with stresses etc that @robbievanleeuwen showed in the example calculation). So overall I have no idea, maybe SPColumn is the issue in your comparison as you guys concluded.

However I get near perfect agreement in some instances, and in others out by some amount that are beyond normal variance due to rounding and precision. It gets weirder I guess, and needs more investigation when I have further time!

**Reply by Agent6-6-6** [reply]
@connorferster, @robbievanleeuwen, I've gotten partly to the bottom of these differences (I think at least):-

- SPColumn makes the assumption that irrespective of the reinforcement layout that the axial load is applied at the location of the gross concrete section centroid (i.e. ignoring reinforcement). This is the same assumption (rightly or wrongly) I've made in developing an independent hand check.
- concreteproperties does not appear to do this. I think looking at the code it assumes the elastic centroid which is inclusive of any offset from the gross concrete sections centroid due to the presence of reinforcement, though finding it a little hard to interpret? 

Obviously with a symmetrical rectangular section, but with unsymmetrical reinforcement this leads to differences in any outputs.

To demonstrate what seems like an error in the concreteproperties approach, if we look at for example when the entire section is in tension and there is no compression at all in the concrete to determine the moment capacity at this point:-

- Consider the following example:- 
  - 800x500 section, 
  - 50 cover to main bars, 
  - 5 x 500MPa 20 and 25 diameter bars in unsymmetrical arrangement at the pure tension control point. 
  - Only the bars are therefore carrying force (c=0), and bars are at yield at the max tension capacity of the section. 
- To maintain force and moment equilibrium, a moment of _147.1 kNm_ **must** be applied. 
- However concreteproperties calculates at the max tension load point a _139.1 kNm_ moment being applied to maintain equilibrium (force and moment), so something is not correct in the concreteproperties calculation regarding the assumption of where the forces are being applied?? 
![image](https://user-images.githubusercontent.com/36808164/193429492-4485c05f-3a6d-4511-b556-f03e852d8e94.png)

I guess it must be remembered if we are comparing an analysis value taken from a centreline bar type element, that the forces we are comparing to are on the geometric centroid of the concrete section.

**Reply by robbievanleeuwen** [reply]
@Agent6-6-6 @connorferster some good discussion here. Hopefully I'll cover everything that hasn't already been solved...

- **Holes in concrete** - I don't see this making a significant difference but happy to be otherwise proven wrong. ``concreteproperties`` only includes concrete where there is concrete and even with the neutral axis within the bar depth should give correct results. For example, if the neutral axis coincides with the bar's centre, the bar will return zero force (force above in compression = force below in tension). If the neutral axis is slightly above the bar's centre, the bar will return a tension force, again even though evaluated at a point, this net force is consistent with summing the compressive areas and tensile areas of the bar so no material is lost or not considered. All materials are modelled consistently and I believe this is more consistant than the approach you describe, but perhaps I've missed something.
- ``n_points`` - I don't think the difference described is appreciable, probably more attributable to numerical differences e.g. neutral axis convergence slightly different between both cases etc.
- **Where to take moments about?** - I believe this is the difference and have come around to taking moments about the centroid of the concrete only. Technically this is arbitrary, but as you say, most users will be getting their moment results from a frame program where they have most likely entered a concrete cross-section and no reinforcement. These moments will be computed about this unreinforced centroid. Hence, for consistency, I agree that the point about which moments should be calculated is the unreinforced cross-section, it is currently the reinforced cross-section in ``concreteproperties``. There's some good discussion on this [here](https://www.eng-tips.com/viewthread.cfm?qid=498525) by @buddyd16 (I think?)
- But this begs the question - what to do for composite steel-concrete sections which would probably be modelled in the frame program? Should the approach be to ignore unmeshed areas (likely bars) in the centroid calculation, but include meshed areas (likely larger sections)? Either way, what is agreed on should also be communicated in the docs.

Once we agree on the above I will convert this to an issue and get it changed.

**Reply by Agent6-6-6** [reply]
> But this begs the question - what to do for composite steel-composite sections which would probably be modelled in the frame program? Should the approach be to ignore unmeshed areas (likely bars) in the centroid calculation, but include meshed areas (likely larger sections)? Either way, what is agreed on should also be communicated in the docs.

I guess if an analysis dictates anything different to the gross concrete centroid, then the design actions would need to be transformed to this location by the user for direct comparison to the concretesection analysis results? So at least you're comparing apples with apples?

Maybe though the composite cross sections needs a bit more thought though. But fundamentally I wonder if it is really any different than an unsymmetrical reinforcement layout. I guess if modelling in a frame analysis there is a possibility of the centroid of the composite section being offset from the concrete centroid, but provided the user understands that the frame analysis output actions can be or need to be transformed in those types of cases to the gross concrete section centroid to allow for direct comparison of analysis actions vs concretesection capacities? 

Potentially you might have the option of by default consider analysis about centroid of the concrete, but have option of specifying the location of the analysis centreline for other more exotic arrangements which would directly address any difference between a frame analysis composite centroid based action and a cross section analysis capacity being resolved to the exact same location?

Maybe returning a warning if meshed reinforcement + concrete centroid is not within some tolerance of the gross concrete centroid, to ensure user takes note and transforms actions appropriately is a good starting point?


I kind of think to go along with this conversation maybe the docs also need a section/page on the analysis assumptions so user is fully informed on any assumptions and possible pitfalls. Most of this is assumed in any analysis, but no harm in explicitly stating it for clarity. You know the usual stuff like for example:-

- Plane sections remain plane (Bernoulli), i.e. sections perpendicular to the axis of bending that are plane before bending remain plane after bending

- The strains in both the concrete and the reinforcement varies linearly through the depth of the member, i.e. cross section strains are proportional to the distance from the neutral axis

- The strain in the reinforcement is equal to the strain in the concrete at the same level/position, i.e. a perfect bond exists between the reinforcement and the concrete

- Any reinforcement is considered as being characterised by a single strain value at the centre of the reinforcement, i.e. for the analysis the strain gradient across the bar diameter is not considered

- The maximum strain at the extreme concrete compression fibre is assumed to reach a maximum limiting compressive strain

- The stresses in the concrete and reinforcement can be determined from the strains using stress-strain curves for both concrete and steel

- The tensile strength of concrete is neglected in strength calculations
- The member does not buckle before the ultimate strength of the cross section is attained

- The compressive stress-strain relationship for concrete may be assumed to be rectangular, trapezoidal, bilinear, parabolic, or any other arbitrary shape that results in prediction of strength in substantial agreement with the results of comprehensive tests
- Other???
 

I was planning on looking at what occurs in the two methods when the concrete is a hole vs considering the effective net stress in the bar as the compression block traversed the bar depth to see if there was any differences or blips in the section capacity achieved. Yet to find time/motivation to do it though!

**Reply by robbievanleeuwen** [reply]
I think I'll go with defaulting to using the gross section centroid (no material properties applied), with the option to use a user defined centroid if desired.

Fully on board with the assumpotions page, will create a new issue! #52

**Comment by Agent6-6-6** [comment]
@connorferster I know spColumn simply considers the reinforcement area as being lumped only at the defined centre of the bar. It does account for the reduced concrete area due to the bar, but the bar itself and forces/stresses are only accounted for based on the setout to the centre. In the past I've compared circular sections using the compression area as a true circular segment and got the exact same answers.
 
So if concreteproperties actually considers the real size and shape of the bar (as it appears to do with the mesh plots, I'm sure @robbievanleeuwen can confirm) then it will probably lead to some subtle differences like you've indicated. 

I'd imagine then as well that considering the bars as being square as per your example @connorferster will also lead to further subtle differences. I've not looked into exactly how concreteproperties analysis is working but from the plots I'm just making an educated guess here that it's leveraging a finite element formulation and not working from a first principles section analysis considering the bars as occurring at just the centre of the bar.

**Reply by robbievanleeuwen** [reply]
@Agent6-6-6, *concreteproperties* does the same in it's calculation. The strains/stresses are evaulated at the centre of the bar and the force is just the area of the bar multiplied by the stress. Because the area of the square bars is equal to the provided area (700 mm^2 in this case) this should not change the results of the calculation, unless the neutral axis passes directly through a bar in which you would expect miniscule differences.

*concreteproperties* uses a very basic finite element formulation to evalute the concrete force based on the provided stress strain profile. Because all stress-strain profiles are piecewise linear this method gives an exact result and is mesh independent (the backend generates a very coarse mesh). In this example a rectangular stress block is used so the stress is constant over the applied area and all the finite element formulation is doing is summing the area of the concrete and multiplying by the concrete stress.

See above thread for a hand calc verifying the *concreteproperties* result.
